File: heap/0378_kth_smallest_element_in_a_sorted_matrix/0378_kth_smallest_element_in_a_sorted_matrix.go
package main 

import (
	"fmt"
)

func kthSmallest(mat [][]int, k int) int {
	n := len(mat)

	lo, hi := mat[0][0], mat[n-1][n-1]

	for lo < hi {
		mid := lo + (hi-lo) >> 2
		count := 0
		j := n-1
		for i := 0; i < n; i++ {
			for j >= 0 && mat[i][j] > mid {
				j--
			}
			count += j + 1
		}

		if count < k {
			lo = mid + 1
		} else {
			hi = mid
		}
	}
	return lo
}

func main() {
	mat := [][]int{{1,5,9},{10,11,13},{12,13,15}}
	k := 8
	res := kthSmallest(mat, k)
	fmt.Println(res)
}

File: heap/0378_kth_smallest_element_in_a_sorted_matrix/0378_kth_smallest_element_in_a_sorted_matrix.py

class Solution(object):
    def kthSmallest(self, matrix, k):
        left = matrix[0][0]
        right = matrix[-1][-1] +1
        while left < right:
            mid = (left + right) / 2
            i = len(matrix) - 1
            j = 0
            count = 0
            while i >= 0 and j < len(matrix[0]):
                if matrix[i][j] <= mid:
                    j += 1
                    count += (i+1)
                else:
                    i -= 1

            if count < k:
                left = mid + 1
            else:
                right = mid
        return left

    
matrix = [[1,5,9],[10,11,13],[12,13,15]]
k = 8
res = Solution().kthSmallest(matrix,k)
print(res)